### PR TITLE
Implement cheaper einsum path optimizer 

### PR DESCRIPTION
**Pull request by fritzo**
Addresses https://github.com/dgasmith/opt_einsum/issues/56
Mirrors upstream https://github.com/dgasmith/opt_einsum/pull/60

 The `opt_einsum` path optimizer is slow for large paths. This PR implements a cheaper path optimizer and switches to use this as the default optimizer in Pyro.

## Performance

- `test_enum.py` is roughly unchanged.
- For `examples/hmm.py` the new optimizer is about 100x faster, bringing path optimization down from 70% of total runtime to <1%. Therefore I've removed our gross `cached_paths` context manager, whose main purpose was to speed up the hmm example.
- `test_elbo_dbn_growth` is fixed by this PR: the new optimizer seems to improve sharing.

## Tested

- [x] exercised by existing tests
- [x] tests for path contraction
- [x] manually checked that hmm example is just as cheap
- [x] ensure optimizer is sufficiently deterministic to allow sharing
- [x] added a notebook to visualize asymptotic growth in HMM and DBN tests

### CODE REVIEW

File: examples/hmm.py [hunk 1]
```diff
@@ -251,37 +250,33 @@ def main(args):
     optim = Adam({'lr': args.learning_rate})
     svi = SVI(model, guide, optim, elbo)
 
-    # We'll speed up parameter tuning by caching the message passing paths
-    # created by opt_einsum. It's always safe to delete this temp file.
-    with cached_paths('data/opt_einsum_path_cache.pkl'):
```
Comment (fritzo) [inline examples/hmm.py]: The only change in this file is removal of this with statement.

**Comment by fritzo** [comment]
@neerajprad if possible I'd like to get this in before release, so we can delete the `cached_paths` context manager.

File: tests/ops/einsum/test_paths.py [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+    logging.debug(u'opt_einsum took {}s:\n{}'.format(opt_time, opt_info))
+    pyro_flops = float(re.search('Optimized FLOP count:(.*)', pyro_info).group(1))
+    opt_flops = float(re.search('Optimized FLOP count:(.*)', opt_info).group(1))
+    assert pyro_flops <= opt_flops * 1.5 + 2.0
+
+    # Check path correctness.
+    try:
+        expected = opt_einsum.contract(equation, *operands, backend='torch', optimize=opt_path)
+    except RuntimeError:
+        return  # ignore torch not implemented errors
```
[the remark this answers is not in the document]
Comment (fritzo) [reply tests/ops/einsum/test_paths.py]: Well these tests are testing path generation, whereas the failure is in executing the paths. I think it's reasonable to say that the path test passes, even if there are errors in downstream usage of those paths.

File: tests/ops/einsum/test_paths.py [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+    'fbb,dfd,fc,fc->',
+    'afd,ba,cc,dc->bf',
+    'adb,bc,fa,cfc->d',
+    'bbd,bda,fc,db->acf',
+    'dba,ead,cad->bce',
+    'aef,fbc,dca->bde',
+]
+
+EQUATIONS += [
+    '->',
```
[the remark this answers is not in the document]
Comment (fritzo) [reply tests/ops/einsum/test_paths.py]: I'm trying to separate the older copyrighted code from our new code.

File: pyro/ops/einsum/__init__.py [hunk 4]
```diff
@@ -1,62 +1,60 @@
...
 
-import contextlib
-import os
-
 import opt_einsum
-from six.moves import cPickle as pickle
+
+from pyro.ops.einsum.paths import optimize
 
 _PATH_CACHE = {}
```
[the remark this answers is not in the document]
Comment (fritzo) [reply pyro/ops/einsum/__init__.py]: Sounds like a reasonable thing to do if we ever see the cache grow too large. Until then it's nice to be able to e.g. share the cache across multiple `SVI.step()` calls and even `TraceEnum_ELBO` instances. Let's see what happens :smile:

File: tests/ops/einsum/test_paths.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (fritzo) [reply tests/ops/einsum/test_paths.py]: Note that if at test gets to this point, it has already passed some weak checks (the `assert`s above), and so it seems draconian to call it a failure. It's really our test scrutiny that is failing.

[the remark this answers is not in the document]
Comment (fritzo) [reply tests/ops/einsum/test_paths.py]: Oh and finally, these tests should pass with the `numpy` backend; the failure is actually in the upstream `torch.einsum` which raises a `RuntimeError` (which should really be a `NotImplementedError`)

[the remark this answers is not in the document]
Comment (neerajprad) [reply tests/ops/einsum/test_paths.py]: Thanks for explaining!

**Pull request merged**